File: javascript/binary_search_spec.js
var should = require('chai').should();
var Algos = require('./binary_search');

var Ary = function() {
  return [1, 2, 3, 4];
};

describe('Binary search algorithms', function(){

  it('lower_bound', function(){
    Algos.lower_bound(Ary(), 3).should.be.equal(2);
    Algos.lower_bound(Ary(), 2.5).should.be.equal(2);
    Algos.lower_bound(Ary(), 1).should.be.equal(0);
    Algos.lower_bound(Ary(), 5).should.be.equal(4);
  });

  it('upper_bound', function(){
    Algos.upper_bound(Ary(), 3).should.be.equal(3);
    Algos.upper_bound(Ary(), 2.5).should.be.equal(2);
    Algos.upper_bound(Ary(), 4).should.be.equal(4);
  });

  it('binary_search', function(){
    Algos.binary_search(Ary(), 3).should.be.equal(true);
    Algos.binary_search(Ary(), 0).should.be.equal(false);
    Algos.binary_search(Ary(), 5).should.be.equal(false);
  });

  it('equal_range', function(){
    Algos.equal_range(Ary(), 3).lower_bound.should.be.equal(2);
    Algos.equal_range(Ary(), 3).upper_bound.should.be.equal(3);
    Algos.equal_range(Ary(), 2.5).lower_bound.should.be.equal(2);
    Algos.equal_range(Ary(), 2.5).upper_bound.should.be.equal(2);
  });
});

File: lua/binary_search_spec.lua
local Algos = require('./binary_search')

local ary = {1, 2, 3, 4}
describe('Binary search algorithms', function()
    
    it('lowerBound', function()
        assert.are.same(Algos.lowerBound(ary, 3), 3)
        assert.are.same(Algos.lowerBound(ary, 2.5), 3)
        assert.are.same(Algos.lowerBound(ary, 1), 1)
        assert.are.same(Algos.lowerBound(ary, 5), 5)
    end)

    it('upperBound', function()
        assert.are.same(Algos.upperBound(ary, 3), 4)
        assert.are.same(Algos.upperBound(ary, 2.5), 3)
        assert.are.same(Algos.upperBound(ary, 4), 5)
    end)

    it('binarySearch', function()
        assert.is_true(Algos.binarySearch(ary, 3))
        assert.is_false(Algos.binarySearch(ary, 0))
        assert.is_false(Algos.binarySearch(ary, 5))
    end)

    it('equalRange', function()
        local lower_a, upper_a = Algos.equalRange(ary, 3)
        assert.are.same(lower_a, 3)
        assert.are.same(upper_a, 4)
        local lower_b, upper_b = Algos.equalRange(ary, 2.5)
        assert.are.same(lower_b, 3)
        assert.are.same(upper_b, 3)
    end)
end)
